src/script.rs

```diff
@@ -0,0 +1,642 @@
+            OP_6 => 0x56,
+            OP_7 => 0x57,
+            OP_8 => 0x58,
+            OP_9 => 0x59,
+            OP_10 => 0x5a,
+            OP_11 => 0x5b,
+            OP_12 => 0x5c,
+            OP_13 => 0x5d,
+            OP_14 => 0x5e,
+            OP_15 => 0x5f,
+            OP_16 => 0x60,
+        }
+    }
+}
+
+impl TryFrom<u8> for PushValue {
+    type Error = ();
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0x00 => Ok(OP_0),
+            0x4c => Ok(OP_PUSHDATA1),
+            0x4d => Ok(OP_PUSHDATA2),
+            0x4e => Ok(OP_PUSHDATA4),
+            0x4f => Ok(OP_1NEGATE),
+            0x50 => Ok(OP_RESERVED),
+            0x51 => Ok(OP_1),
+            0x52 => Ok(OP_2),
+            0x53 => Ok(OP_3),
+            0x54 => Ok(OP_4),
+            0x55 => Ok(OP_5),
+            0x56 => Ok(OP_6),
+            0x57 => Ok(OP_7),
+            0x58 => Ok(OP_8),
+            0x59 => Ok(OP_9),
+            0x5a => Ok(OP_10),
+            0x5b => Ok(OP_11),
+            0x5c => Ok(OP_12),
+            0x5d => Ok(OP_13),
+            0x5e => Ok(OP_14),
+            0x5f => Ok(OP_15),
+            0x60 => Ok(OP_16),
+            _ => {
+                if value <= 0x60 {
+                    Ok(PushdataBytelength(value))
+                } else {
+                    Err(())
+                }
+            }
+        }
+    }
+}
+
+impl From<Operation> for u8 {
+    fn from(value: Operation) -> Self {
+        // This is how you get the discriminant, but using `as` everywhere is too much code smell
+        value as u8
+    }
+}
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct ScriptNum(i64);
+
+impl ScriptNum {
+    const DEFAULT_MAX_NUM_SIZE: usize = 4;
+
+    pub fn new(
+        vch: &Vec<u8>,
+        require_minimal: bool,
+        max_num_size: Option<usize>,
+    ) -> Result<Self, ScriptNumError> {
+        let max_num_size = max_num_size.unwrap_or(Self::DEFAULT_MAX_NUM_SIZE);
+        if vch.len() > max_num_size {
+            return Err(ScriptNumError::Overflow {
+                max_num_size,
+                actual: vch.len(),
+            });
+        }
+        if require_minimal && !vch.is_empty() {
+            // Check that the number is encoded with the minimum possible
+            // number of bytes.
+            //
+            // If the most-significant-byte - excluding the sign bit - is zero
+            // then we're not minimal. Note how this test also rejects the
+            // negative-zero encoding, 0x80.
+            if (vch.last().unwrap_or_else(|| unreachable!()) & 0x7F) == 0 {
+                // One exception: if there's more than one byte and the most
+                // significant bit of the second-most-significant-byte is set
+                // it would conflict with the sign bit. An example of this case
+                // is +-255, which encode to 0xff00 and 0xff80 respectively.
+                // (big-endian).
```

Comment:
It's confusing to say "big-endian" here, since the elements are not 16-bit integers.
```suggestion
                // is +-255, which encode to [0xff, 0x00] and [0xff, 0x80] respectively.
```